asf_tools/composite.py

```diff
@@ -0,0 +1,308 @@
+            if my_zone != home_zone:
+                logging.info(f"Reprojecting {fi} to {name}")
+                if hemi == "N":
+                    proj = ('EPSG:326%02d' % int(home_zone))
+                else:
+                    proj = ('EPSG:327%02d' % int(home_zone))
+                gdal.Warp(name, fi, dstSRS=proj, xRes=pix_size, yRes=pix_size, targetAlignedPixels=True)
+                gdal.Warp(aname, afi, dstSRS=proj, xRes=pix_size, yRes=pix_size, targetAlignedPixels=True)
+                new_files.append(name)
+            elif x < pix_size:
+                # Need to reproject to desired resolution
+                logging.info(f"Changing resolution of {fi} to {pix_size}")
+                gdal.Warp(name, fi, xRes=pix_size, yRes=pix_size, targetAlignedPixels=True)
+                gdal.Warp(aname, afi, xRes=pix_size, yRes=pix_size, targetAlignedPixels=True)
+                new_files.append(name)
+            else:
+                logging.info(f"No reprojection needed; Linking {fi} to {name}")
+                os.symlink(fi, name)
+                os.symlink(afi, aname)
+                new_files.append(name)
+        else:
+            logging.info(f"Found previous reproj file {fi} - taking no action")
+            new_files.append(fi)
+
+    logging.info("All files completed")
+    return new_files
+
+
+def make_composite(outfile, infiles=None, path=None, pol=None, resolution=None):
+
+    '''Create a composite mosaic of infiles using inverse area weighting to adjust backscatter'''
+
+    logging.info(f"make_composite: {outfile} {infiles} {path} {pol} {resolution}")
+    if pol is None:
+        pol = "VV"
+
+    # Establish input file list
+    if path:
+        logging.info("Searching for list of files to process")
+
+        # New format directory names
+        infiles_new = glob.glob(os.path.join(path, f"20*/S1?_IW_*RTC*/*{pol}.tif"))
+
+        # Old format diretory names
+        infiles_old = glob.glob(os.path.join(path, f"20*/PRODUCT/*{pol}.tif"))
+
+        infiles = infiles_new
+        infiles.extend(infiles_old)
+        cnt = len(infiles)
+        logging.info(f"Found {cnt} files to process")
+    else:
+        cnt = len(infiles)
+        logging.info("Found list of {cnt} input files to process")
+    infiles.sort()
+    logging.debug(f"Input files: {infiles}")
+
+    # resample infiles to maximum resolution & common UTM zone
+    resampled_files = reproject_to_median_utm(infiles, pol, resolution=resolution)
+    if len(resampled_files) == 0:
+        Exception("Unable to resample files")
+
+    # Get pixel size
+    x, y, trans, proj = saa.read_gdal_file_geo(saa.open_gdal_file(resampled_files[0]))
+    pixel_size_x = trans[1]
+    pixel_size_y = trans[5] 
+    logging.info(f"{resampled_files[0]} x = {pixel_size_x} y = {pixel_size_y}")
+
+    # Get extent of union of all images
+    extents = []
+    for fi in resampled_files:
+        ulx, lrx, lry, uly = saa.getCorners(fi)
```

Comment:
This work, and the following loop through the extents could be done within the same loop

Reply:
I must be missing something again, because this loop sets up the extents of the whole output file.  I use the sizes from this loop to create the arrays outputs, weights, and counts.  If I don't run through the loop first to determine the size of these arrays, then how do we size our work arrays correctly?

Reply:
Right -- calculating the full extent is needed ahead of time.
 
The work I was talking about is opening the files and reading them in. They don't need to be opened here to get the corners, transformation, or projection (gdal info lets us get that directly) so we can read the files in the second loop and it's clearer were looking through the files.

Reply:
Here, the call is to read_gdal_file__geo_, which only reads the geo information, not the data.  So, I don't think it's an issue.  I mean, presumably in order to read the geo information the file has to be opened - whether gdal does the opening of the file or we do it, it still has to be done.  Data from the files is only read a single time, that being in the main loop of the code (loop over extents).